### PR TITLE
feat(profiling): use feature options to handle the ingestion of un-sampled profiles

### PR DESCRIPTION
**Pull request by viglia**
This will let us use data of non sampled profiles to enhance current aggregation (slowest function) with more data.

### CODE REVIEW

**Review by Zylphrex — APPROVED** [review]

**Review by Zylphrex — CHANGES_REQUESTED** [review]

**Review by jjbayer — APPROVED** [review]
Agree with @Zylphrex 's comment, the rest looks good to me!

File: src/sentry/profiles/task.py [hunk 1]
```diff
@@ -62,13 +62,20 @@ def process_profile_task(
         message_dict = msgpack.unpackb(payload, use_list=False)
         profile = json.loads(message_dict["payload"], use_rapid_json=True)
 
+        if (
+            not message_dict.get("sampled", True)
+            and options.get("profiling.profile_metrics.unsampled_profiles.sample_rate") == 0.0
```
Comment (Zylphrex) [inline src/sentry/profiles/task.py:67]: did you mean to compare the sample rate to a random number here? Also, why do you read the option in the consumer as well as in the celery task?

File: src/sentry/profiles/consumers/process/factory.py [hunk 2]
```diff
@@ -1,17 +1,26 @@
...
 
+from sentry import options
 from sentry.processing.backpressure.arroyo import HealthChecker, create_backpressure_step
 from sentry.profiles.task import process_profile_task
 
 
 def process_message(message: Message[KafkaPayload]) -> None:
-    process_profile_task.s(payload=message.payload.value).apply_async()
+    msg_payload = message.payload.value
+    message_dict = msgpack.unpackb(msg_payload, use_list=False)
```
Comment (Zylphrex) [inline src/sentry/profiles/consumers/process/factory.py:18]: For the first round, only profiles from select organizations will make it here from relay, right? Adding the message pack in this step can significantly increase the CPU needs of the consumer so we should be mindful of that and monitor it.

**Review by phacops — CHANGES_REQUESTED** [review]
Can we put the `sampled` value in the Kafka headers instead? The consumer can then check on it or pass it directly to the Celery task as an argument and let the Celery task deal with the profile.

Removing deserialization is the best thing that happened to this consumer and I want to avoid adding it back to check a boolean.

File: src/sentry/profiles/consumers/process/factory.py (hunk 2 above)
Comment (viglia) [reply src/sentry/profiles/consumers/process/factory.py:17]: yep, I've double-checked with Pierre a few months back when I first opened this PR to make sure if this was ok.

**Comment by viglia** [comment]
> Can we put the `sampled` value in the Kafka headers instead? The consumer can then check on it or pass it directly to the Celery task as an argument and let the Celery task deal with the profile.
> 
> Removing deserialization is the best thing that happened to this consumer and I want to avoid adding it back to check a boolean.

@phacops that's fine by me, but I remember double-checking with you several months ago and it was fine back then.

If we can skip the deserialization better though 👍🏻

**Comment by phacops** [comment]
> @phacops that's fine by me, but I remember double-checking with you several months ago and it was fine back then.

Right, I said that but now I think we have a better way to do it with the headers.

File: src/sentry/profiles/consumers/process/factory.py [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
 
+from sentry import options
 from sentry.processing.backpressure.arroyo import HealthChecker, create_backpressure_step
 from sentry.profiles.task import process_profile_task
 
 
 def process_message(message: Message[KafkaPayload]) -> None:
-    process_profile_task.s(payload=message.payload.value).apply_async()
+    msg_payload = message.payload.value
+    msg_headers = get_headers_dict(message.payload.headers)
```
Comment (phacops) [inline src/sentry/profiles/consumers/process/factory.py:16]: Why don't you just try to get the field you need?

Comment (viglia) [reply src/sentry/profiles/consumers/process/factory.py:16]: @phacops nope, `Message` headers values are `String`.

Comment (phacops) [reply src/sentry/profiles/consumers/process/factory.py:16]: Got it. I'd still zeroed in on the field you're looking for and not convert the whole dict.

Comment (viglia) [reply src/sentry/profiles/consumers/process/factory.py:16]: other examples from existing `Sentry` codebase: https://github.com/getsentry/sentry/blob/master/src/sentry/eventstream/kafka/protocol.py#L168-L169

File: src/sentry/profiles/consumers/process/factory.py [hunk 4]
```diff
@@ -1,17 +1,23 @@
...
 from arroyo.types import Commit, Message, Partition
 
+from sentry import options
 from sentry.processing.backpressure.arroyo import HealthChecker, create_backpressure_step
 from sentry.profiles.task import process_profile_task
 
 
 def process_message(message: Message[KafkaPayload]) -> None:
-    process_profile_task.s(payload=message.payload.value).apply_async()
+    msg_payload = message.payload.value
```
Comment (phacops) [inline src/sentry/profiles/consumers/process/factory.py:15]: You don't need this variable. It's either not used or only used once.

File: src/sentry/profiles/consumers/process/factory.py [hunk 5]
```diff
@@ -32,3 +38,12 @@ def create_with_partitions(
             health_checker=self.health_checker,
             next_step=next_step,
         )
+
+
+def get_headers_dict(headers) -> Dict[str, str]:
+    h = dict()
+    for k, v in headers:
```
Comment (phacops) [inline src/sentry/profiles/consumers/process/factory.py:45]: Are the headers a list of tuples or do you need `items()` here?

Comment (viglia) [reply src/sentry/profiles/consumers/process/factory.py:43]: yep, list of tuples: https://github.com/getsentry/arroyo/blob/0b4126f52509dc823e592d076747b34febcb3799/arroyo/backends/kafka/consumer.py#L63

File: src/sentry/profiles/consumers/process/factory.py [hunk 6]
```diff
@@ -32,3 +38,12 @@ def create_with_partitions(
...
+        if isinstance(v, bytes):
+            v = str(v, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (viglia) [reply src/sentry/profiles/consumers/process/factory.py]:
```
>>> str(b"true")
"b'true'"
>>> str(b"true", encoding="utf-8")
'true'
```

File: src/sentry/profiles/consumers/process/factory.py [hunk 7]
```diff
@@ -32,3 +36,11 @@ def create_with_partitions(
             health_checker=self.health_checker,
             next_step=next_step,
         )
+
+
+def is_sampled(headers) -> bool:
+    for k, v in headers:
+        if k == "sampled":
+            if isinstance(v, bytes):
+                return str(v, encoding="utf-8") == "true"
```
Comment (Zylphrex) [inline src/sentry/profiles/consumers/process/factory.py:45]: `v.decode("utf-8")` is probably what you're looking for here

Comment (viglia) [reply src/sentry/profiles/consumers/process/factory.py:45]: @Zylphrex both works actually. Is there a substantial difference to opt for the decode method instead?

File: src/sentry/profiles/task.py [hunk 8]
```diff
@@ -56,19 +56,24 @@ class VroomTimeout(Exception):
...
+    sampled: bool = True,
     **kwargs: Any,
 ) -> None:
     if payload:
         message_dict = msgpack.unpackb(payload, use_list=False)
         profile = json.loads(message_dict["payload"], use_rapid_json=True)
 
+        if not sampled and not options.get("profiling.profile_metrics.unsampled_profiles.enabled"):
+            return
+
```
Comment (phacops) [inline src/sentry/profiles/task.py:68]: You can move that before `if payload`.

File: src/sentry/profiles/consumers/process/factory.py (hunk 7 above)
Comment (Zylphrex) [reply src/sentry/profiles/consumers/process/factory.py:45]: https://docs.python.org/3/library/stdtypes.html#str

Don't think there's a huge difference but the docs mention

> Passing a [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) object to [str()](https://docs.python.org/3/library/stdtypes.html#str) without the encoding or errors arguments falls under the first case of returning the informal string representation (see also the [-b](https://docs.python.org/3/using/cmdline.html#cmdoption-b) command-line option to Python).

which is why you need to specify the encoding here. This is just an uncommon usage of `str` that I don't think we have anywhere else. So the suggestion is more for style.

**Review by phacops — APPROVED** [review]

**Review by Zylphrex — APPROVED** [review]

**Pull request merged**